Domain decomposition example using LDC

ktangsali opened the pull request:
<!-- markdownlint-disable MD013-->
# Modulus Pull Request

## Description
A simple example showing how to use domain decomposition using X-PINN style and FBPINN style in Modulus. The concept is demonstrated on a Lid-Driven Cavity case.

Domain is divided as shown below:

![image](https://github.com/NVIDIA/modulus-sym/assets/71059996/d2136854-9393-43a4-9e43-8db6199a576d)

For FBPINN, the basis functions are used as below:

![image](https://github.com/NVIDIA/modulus-sym/assets/71059996/9936d9db-ba59-4c96-8a70-08d39d7d4a25)


Results:

![image](https://github.com/NVIDIA/modulus-sym/assets/71059996/f3718cc3-ffda-42d6-9971-bb68b3e3df19)


## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/modulus-sym/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [x] The [CHANGELOG.md](https://github.com/NVIDIA/modulus-sym/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/modulus-sym/issues) is linked to this pull request.

## Dependencies

<!-- Call out any new dependencies needed if any -->

ktangsali commented:
/blossom-ci

mnabian commented:
@ktangsali in my opinion this example should demonstrate how the training workload for each subdomain can be distributed on different GPUs, which does not exist currently in your example.
Also, if for any reason, you are interested in single GPU training with multiple domains, consider the FBPINNs approach, which essentially helps you get rid of the interface constraints. Your example can be easily extended to FBPINNs by removing the interface constraints, and switching from Heaviside to a smooth function (e.g. hyperbolic tangent) which overlaps with the opposite domain.

mnabian commented:
Because the distributed training component is missing here, should we call this example something other than XPINNs? Maybe domain decomposition?

ktangsali commented on examples/ldc/ldc_2d_xpinn.py:
```diff
@@ -0,0 +1,270 @@
+    from sympy import Piecewise, Heaviside, tanh
+    from modulus.sym.node import Node
+
+    # nodes for interface condition. Currently only dirichlet, but can be extended to neumann too
```
Thanks! Yes, definitely it is possible to add that, however for this example, just the solution continuity suffices so I have added only that. I have also added comments to reflect the same.

ktangsali commented:
Thanks @mnabian , I have addressed your comments and also renamed the example. I think these two examples are a great demonstration of just how to use different neural networks in different domains of the problem and couple them via interface/basis functions. The distributed part is a nice-to-have and we can add that in a separate PR.

ktangsali commented:
/blossom-ci
